### PR TITLE
efl_ui_spin: add overall tests for stabilization of API

### PR DESCRIPTION
**Pull request by marcelhollerbach**
This also adds tests and fixes to simular classes or inheriting classes. This brings efl_ui_spin closer to beeing stable.

### CODE REVIEW

File: src/lib/elementary/efl_ui_progressbar.c [hunk 1]
```diff
@@ -593,6 +593,17 @@ _progressbar_part_value_get(Efl_Ui_Progressbar_Data *sd, const char* part)
...
 {
+   if (val < sd->val_min)
+     {
+        ERR("Error, value is lower than minimum");
+        return;
+     }
+
+   if (val > sd->val_max)
+     {
+        ERR("Error, value is bigger than minimum");
```
Comment (segfaultxavi) [inline src/lib/elementary/efl_ui_progressbar.c]:
```suggestion
        ERR("Error, value is bigger than maximum");
```

File: src/lib/elementary/efl_ui_progressbar.c [hunk 2]
```diff
@@ -644,6 +655,16 @@ _efl_ui_progressbar_pulse_get(const Eo *obj EINA_UNUSED, Efl_Ui_Progressbar_Data
...
 _efl_ui_progressbar_efl_ui_range_display_range_limits_set(Eo *obj, Efl_Ui_Progressbar_Data *sd, double min, double max)
 {
+   if (max < min)
+     {
+        ERR("Wrong params. min(%lf) is bigger than max(%lf).", min, max);
+        return;
+     }
+   if (EINA_DBL_EQ(max, min))
+     {
+        ERR("min and max are of the same value. They must have a distance from each other!");
```
Comment (segfaultxavi) [inline src/lib/elementary/efl_ui_progressbar.c]:
```suggestion
        ERR("min and max must have a different value");
```

File: src/lib/elementary/efl_ui_slider.c [hunk 3]
```diff
@@ -719,7 +719,17 @@ _efl_ui_slider_efl_ui_layout_orientable_orientation_get(const Eo *obj EINA_UNUSE
...
 {
-   if ((sd->val_min == min) && (sd->val_max == max)) return;
+   if (max < min)
+     {
+        ERR("Wrong params. min(%lf) is bigger than max(%lf).", min, max);
+        return;
+     }
+   if (EINA_DBL_EQ(max, min))
+     {
+        ERR("min and max are of the same value. They must have a distance from each other!");
```
Comment (segfaultxavi) [inline src/lib/elementary/efl_ui_slider.c]:
```suggestion
        ERR("min and max must have a different value");
```

File: src/lib/elementary/efl_ui_slider.c [hunk 4]
```diff
@@ -738,7 +748,19 @@ _efl_ui_slider_efl_ui_range_display_range_limits_get(const Eo *obj EINA_UNUSED,
...
-   if (sd->val == val) return;
+   if (val < sd->val_min)
+     {
+        ERR("Error, value is lower than minimum");
+        return;
+     }
+
+   if (val > sd->val_max)
+     {
+        ERR("Error, value is bigger than minimum");
```
Comment (segfaultxavi) [inline src/lib/elementary/efl_ui_slider.c]:
```suggestion
        ERR("Error, value is bigger than maximum");
```

File: src/lib/elementary/efl_ui_spin.c [hunk 5]
```diff
@@ -108,10 +108,13 @@ _efl_ui_spin_efl_ui_range_display_range_limits_set(Eo *obj, Efl_Ui_Spin_Data *sd
...
-        ERR("Wrong params. min(%lf) is bigger than max(%lf). It will swaped.", min, max);
-        double t = min;
-        min = max;
-        max = t;
+        ERR("Wrong params. min(%lf) is bigger than max(%lf).", min, max);
+        return;
+     }
+   if (EINA_DBL_EQ(max, min))
+     {
+        ERR("min and max are of the same value. They must have a distance from each other!");
```
Comment (segfaultxavi) [inline src/lib/elementary/efl_ui_spin.c]:
```suggestion
        ERR("min and max must have a different value");
```

File: src/lib/elementary/efl_ui_spin.c [hunk 6]
```diff
@@ -153,9 +156,16 @@ EOLIAN static void
...
-   else if (val > sd->val_max)
-     val = sd->val_max;
+     {
+        ERR("Error, value is lower than minimum");
+        return;
+     }
+
+   if (val > sd->val_max)
+     {
+        ERR("Error, value is bigger than minimum");
```
Comment (segfaultxavi) [inline src/lib/elementary/efl_ui_spin.c]:
```suggestion
        ERR("Error, value is bigger than maximum");
```

File: src/tests/elementary/efl_ui_test_spin.c [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+   ck_assert_int_eq(changed, EINA_TRUE);
+   ck_assert_int_eq(min_reached, EINA_TRUE);
+   ck_assert_int_eq(max_reached, EINA_FALSE);
+   changed = EINA_FALSE;
+   min_reached = EINA_FALSE;
+   max_reached = EINA_FALSE;
+}
+EFL_END_TEST
+
+EFL_START_TEST (spin_wheel_test)
```
Comment (zmike) [inline src/tests/elementary/efl_ui_test_spin.c]: I'd like to see this test also verify that callbacks are triggered.

File: src/tests/elementary/efl_ui_test_spin.c [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+{
+   Eina_Bool changed = EINA_FALSE, min_reached = EINA_FALSE, max_reached = EINA_FALSE;
+
+   efl_ui_range_limits_set(spin, -100.0, 100.0);
+   efl_ui_range_value_set(spin, 0.0);
+   efl_ui_range_step_set(spin, 10.0);
+   efl_event_callback_add(spin, EFL_UI_SPIN_EVENT_CHANGED, _set_flag, &changed);
+   efl_event_callback_add(spin, EFL_UI_SPIN_EVENT_MIN_REACHED, _set_flag, &min_reached);
```
Comment (zmike) [inline src/tests/elementary/efl_ui_test_spin.c]: If these min/max callbacks are only supposed to trigger once then they need to assert that they are only triggered exactly one time. Same for above test.

**Pull request closed without merging**